### PR TITLE
Support Iceberg refs system table

### PR DESCRIPTION
**Pull request by jackye1995**
## Description

Support refs system table in Iceberg

## Additional context and related issues

Users can query the system table with:

```
SELECT * FROM "table$refs"
```

The schema is the same as the one in Iceberg repo: https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/RefsTable.java

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
(x) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RefsTable.java
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RefsTable.java:80]: Nitpick:
```suggestion
        icebergTable.refs().forEach((refName, ref) -> {
```

**Review by alexjo2144 — COMMENTED** [review]
How do you feel about calling the table `REFERENCES` spelled out?

This also needs to be added to `iceberg.rst` for the docs here: https://trino.io/docs/current/connector/iceberg.html#metadata-tables

**Comment by jackye1995** [comment]
> How do you feel about calling the table REFERENCES spelled out?

I am using `REFS` because that is what Iceberg core library uses, which means Spark and Flink system tables are also names as that. But if you think REFERENCES makes more sense in Trino, happy to name it that way.

**Comment by findinpath** [comment]
> This also needs to be added to iceberg.rst for the docs here: https://trino.io/docs/current/connector/iceberg.html#metadata-tables

See for reference: https://github.com/apache/iceberg/blob/master/format/spec.md#snapshot-reference

**Comment by jackye1995** [comment]
@alexjo2144 updated everything except the name `REFS`, let me know if I need to name it to `REFERENCES`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RefsTable.java
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RefsTable.java:93]: You could add in `PageListBuilder` new methods:

- `appendInteger(Integer)`
- `appendBigint(Long)`

to encapsulate the `null` check and make the code in this class more concise.

**Comment by alexjo2144** [comment]
> > How do you feel about calling the table REFERENCES spelled out?
> 
> I am using `REFS` because that is what Iceberg core library uses, which means Spark and Flink system tables are also names as that. But if you think REFERENCES makes more sense in Trino, happy to name it that way.

@ebyhr or @findepi either of you have thoughts on this? Matching the other engines vs spelling the whole word out.

File: docs/src/main/sphinx/connector/iceberg.rst
Comment (alexjo2144) [inline docs/src/main/sphinx/connector/iceberg.rst:1250]: I think these are supposed to match width. @mosabua might want to take a look at the docs addition, but thanks for adding it

**Review by alexjo2144 — APPROVED** [review]
Looks good to me, pending input from maintainers on the table name.

File: docs/src/main/sphinx/connector/iceberg.rst
Comment (jackye1995) [reply docs/src/main/sphinx/connector/iceberg.rst]: Updated to match the width

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RefsTable.java
Comment (jackye1995) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RefsTable.java]: Added new methods as suggested

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/util/PageListBuilder.java
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/util/PageListBuilder.java:97]: We could add a helper method likes `SnapshotsTable#checkNonNull` to avoid repeated null check.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RefsTable.java
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RefsTable.java:80]: Just curious, do you know the order of `refs()` results? Is it deterministic?

**Review by ebyhr — APPROVED** [review]

**Comment by bitsondatadev** [comment]
@jackye1995, would you mind replying to the last comments here please?

@ebyhr %comments, is this ready to be merged?

**Comment by jackye1995** [comment]
@ebyhr sorry did not notice you have one more comment, updated. Let me know if there is anything else needed.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RefsTable.java
Comment (jackye1995) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RefsTable.java:81]: it's not, like the table properties metadata table, it's determined by the order from the hash map.

**Comment by bitsondatadev** [comment]
> @ebyhr sorry did not notice you have one more comment, updated. Let me know if there is anything else needed.

Thanks @jackye1995! I think there are a few other small naming fixes in #15646 if you can address those as well please!

After that @ebyhr should take a look shortly! You rock!

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/util/PageListBuilder.java
Comment (jackye1995) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/util/PageListBuilder.java:97]: Updated, also updated related methods in `SnapshotsTable`. Let me know if this is what you imagine

**Comment by electrum** [comment]
I think it's fine to match the other implementations, especially since the intent here is to have the same schema.

**Comment by jackye1995** [comment]
@electrum @ebyhr @findepi is there any plan to merge this and the related PRs for Iceberg reference support? We are about to cut Iceberg 1.2 branch, and have achieved all the same supports in Spark and Flink, would be great to have feature parity in Trino.

**Comment by bitsondatadev** [comment]
@electrum

**Comment by ebyhr** [comment]
Could you rebase on master to resolve conflicts?

**Comment by jackye1995** [comment]
Can I get a confirmation if this PR is ever gonna be merged once rebased? If not I will just go ahead to apply the patch internally and you can close this PR instead of me continuing to rebase.

**Comment by bitsondatadev** [comment]
@jackye1995  I'm meeting with @electrum tomorrow I'll ask him to respond. I really want this one land!

**Comment by bitsondatadev** [comment]
@jackye1995 would you also be able to do this change suggested by David to go back to `REFS` instead of `REFERENCES`?

https://trinodb.slack.com/archives/CP1MUNEUX/p1678319193961979?thread_ts=1677280433.219989&channel=CP1MUNEUX&message_ts=1678319193.961979


I will also update you on what David says about how close this is to merge.

**Comment by ebyhr** [comment]
> would you also be able to do this change suggested by David to go back to REFS instead of REFERENCES?

The current implementation is already `refs`.

**Comment by bitsondatadev** [comment]
> > would you also be able to do this change suggested by David to go back to REFS instead of REFERENCES?
> 
> 
> 
> The current implementation is already `refs`. 

Ooops 🫣🫠

**Comment by jackye1995** [comment]
Close PR, this will be available in AWS Athena by 3/31 in all AWS regions.

**Review by electrum — APPROVED** [review]

**Comment by bitsondatadev** [comment]
@electrum looks like the tests passed! Are we ready to merge this?

**Comment by electrum** [comment]
There was another conflict. Rebased again.

**Comment by findepi** [comment]
Per @ebyhr 's https://github.com/trinodb/trino/pull/16576 it seems the PR broke master builds.
@jackye1995 can you please help us and remind to `/test-with-secrets`?

**Pull request merged**
